File: app/src/main/java/com/example/terabox/MainActivity.java
package com.example.terabox;

import android.content.Intent;
import android.graphics.Typeface;
import android.os.Bundle;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextPaint;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.text.style.StyleSpan;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.example.terabox.databinding.ActivityLoginBinding;

public class MainActivity extends AppCompatActivity {

    private ActivityLoginBinding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityLoginBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        setupLegalText();
        View.OnClickListener goDashboard = v -> openDashboard();

        binding.buttonApple.setOnClickListener(goDashboard);
        binding.buttonGoogle.setOnClickListener(goDashboard);
        binding.buttonFacebook.setOnClickListener(goDashboard);
        binding.buttonEmail.setOnClickListener(goDashboard);
        binding.buttonPhone.setOnClickListener(goDashboard);

        binding.buttonHelp.setOnClickListener(v ->
                Toast.makeText(this, R.string.content_desc_help, Toast.LENGTH_SHORT).show());
    }

    private void setupLegalText() {
        String full = getString(R.string.login_legal);
        String terms = getString(R.string.terms_of_service);
        String privacy = getString(R.string.privacy_policy);
        SpannableString ss = new SpannableString(full);

        int termsStart = full.indexOf(terms);
        int privacyStart = full.indexOf(privacy);
        int linkColor = ContextCompat.getColor(this, R.color.terabox_link);

        if (termsStart >= 0) {
            ss.setSpan(new StyleSpan(Typeface.BOLD), termsStart, termsStart + terms.length(),
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            ss.setSpan(new LinkSpan(linkColor, () -> showStub(terms)), termsStart,
                    termsStart + terms.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        }
        if (privacyStart >= 0) {
            ss.setSpan(new StyleSpan(Typeface.BOLD), privacyStart, privacyStart + privacy.length(),
                    Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            ss.setSpan(new LinkSpan(linkColor, () -> showStub(privacy)), privacyStart,
                    privacyStart + privacy.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        }

        binding.textLegal.setText(ss);
        binding.textLegal.setMovementMethod(LinkMovementMethod.getInstance());
        binding.textLegal.setHighlightColor(android.graphics.Color.TRANSPARENT);
    }

    private void showStub(String label) {
        Toast.makeText(this, label, Toast.LENGTH_SHORT).show();
    }

    private void openDashboard() {
        startActivity(new Intent(this, DashboardActivity.class));
    }

    private static final class LinkSpan extends ClickableSpan {
        private final int color;
        private final Runnable action;

        LinkSpan(int color, Runnable action) {
            this.color = color;
            this.action = action;
        }

        @Override
        public void updateDrawState(@NonNull TextPaint ds) {
            super.updateDrawState(ds);
            ds.setColor(color);
            ds.setUnderlineText(true);
            ds.bgColor = android.graphics.Color.TRANSPARENT;
        }

        @Override
        public void onClick(@NonNull View widget) {
            action.run();
        }
    }
}

File: app/src/main/kotlin/com/example/terabox/MainActivity.kt
package com.example.terabox

import android.content.Intent
import android.graphics.Color
import android.graphics.Typeface
import android.os.Bundle
import android.text.SpannableString
import android.text.Spanned
import android.text.TextPaint
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.text.style.StyleSpan
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.example.terabox.databinding.ActivityLoginBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityLoginBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupLegalText()

        val goDashboard = View.OnClickListener { openDashboard() }
        binding.buttonApple.setOnClickListener(goDashboard)
        binding.buttonGoogle.setOnClickListener(goDashboard)
        binding.buttonFacebook.setOnClickListener(goDashboard)
        binding.buttonEmail.setOnClickListener(goDashboard)
        binding.buttonPhone.setOnClickListener(goDashboard)

        binding.buttonHelp.setOnClickListener {
            Toast.makeText(this, R.string.content_desc_help, Toast.LENGTH_SHORT).show()
        }
    }

    private fun setupLegalText() {
        val full = getString(R.string.login_legal)
        val terms = getString(R.string.terms_of_service)
        val privacy = getString(R.string.privacy_policy)
        val ss = SpannableString(full)
        val termsStart = full.indexOf(terms)
        val privacyStart = full.indexOf(privacy)
        val linkColor = ContextCompat.getColor(this, R.color.terabox_link)

        if (termsStart >= 0) {
            ss.setSpan(
                StyleSpan(Typeface.BOLD),
                termsStart,
                termsStart + terms.length,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            ss.setSpan(
                LinkSpan(linkColor) { showStub(terms) },
                termsStart,
                termsStart + terms.length,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
        }
        if (privacyStart >= 0) {
            ss.setSpan(
                StyleSpan(Typeface.BOLD),
                privacyStart,
                privacyStart + privacy.length,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            ss.setSpan(
                LinkSpan(linkColor) { showStub(privacy) },
                privacyStart,
                privacyStart + privacy.length,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
        }

        binding.textLegal.text = ss
        binding.textLegal.movementMethod = LinkMovementMethod.getInstance()
        binding.textLegal.highlightColor = Color.TRANSPARENT
    }

    private fun showStub(label: String) {
        Toast.makeText(this, label, Toast.LENGTH_SHORT).show()
    }

    private fun openDashboard() {
        startActivity(Intent(this, DashboardActivity::class.java))
    }

    private class LinkSpan(
        private val color: Int,
        private val action: () -> Unit
    ) : ClickableSpan() {
        override fun updateDrawState(ds: TextPaint) {
            super.updateDrawState(ds)
            ds.color = color
            ds.isUnderlineText = true
            ds.bgColor = Color.TRANSPARENT
        }

        override fun onClick(widget: View) {
            action()
        }
    }
}
